### ISSUE TITLE
[JENKINS-26048] Jenkins no longer cleaning up child processes when build stopped - as of 1.587

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-14#12068]
<p>Similar to bug 22641, Jenkins no longer cleaning up child processes when a job is stopped. </p>

<p>To reproduce : Start any job with a surefire execution, once the tests start running, stop the job through the jenkins interface. The surefire process continues to run. </p>

<p>Just like bug 22641, this behavior started in 1.553, was resolved in 1.565.2, then started again in 1.587. </p>

<p>Rating this as critical because processes build up eventually causing issues with running out of memory or hitting the nproc limit. </p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dbogardus">dbogardus</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-26048" target="_blank">Jenkins no longer cleaning up child processes when build stopped - as of 1.587</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=pi0tras">pi0tras</a>
<li><b>status</b>: Reopened
<li><b>priority</b>: Critical
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 39
<li><b>watchers</b>: 38
<li><b>imported</b>: 2025-11-14
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Similar to bug 22641, Jenkins no longer cleaning up child processes when a job is stopped. </p>

<p>To reproduce : Start any job with a surefire execution, once the tests start running, stop the job through the jenkins interface. The surefire process continues to run. </p>

<p>Just like bug 22641, this behavior started in 1.553, was resolved in 1.565.2, then started again in 1.587. </p>

<p>Rating this as critical because processes build up eventually causing issues with running out of memory or hitting the nproc limit. </p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
CentOS 6, JRE 7 or 8 <br/>
Windows Server 2008 R2, JRE 7 or 8<br/>
```
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=217651&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-217651">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Is this specific to the Maven project type? Or to freestyle?</p>
</pre>
</details>

</details>
<p>Is this specific to the Maven project type? Or to freestyle?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dbogardus's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=217762&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-217762">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Only tested it with Maven projects. </p>
</pre>
</details>

</details>
<p>Only tested it with Maven projects. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=217956&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-217956">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>When you write it started in 1.587, do you mean that it worked in 1.586? Or do you not know when between 1.565.2 and 1.587 it started?</p>
</pre>
</details>

</details>
<p>When you write it started in 1.587, do you mean that it worked in 1.586? Or do you not know when between 1.565.2 and 1.587 it started?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/grayaii.png" title="grayaii's avatar" /> grayaii's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=247727&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-247727">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>We get bitten by this too. A bunch of our jobs have a pre build step that basically does a:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-none">ps aux | grep [the-thing-you-want-to-kill] | grep -v grep | awk "{ print \$2 }"
</pre>
</div></div>

<p>and pass those pids to kill -9</p>

<p>That kills every "zombie/parent-less" process BEFORE the build runs.  This ensures the build will run in a relatively clean environment.
Just be careful what you kill <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>This is only a workaround, but it works for us.
Hope this helps!</p>
</pre>
</details>

</details>
<p>We get bitten by this too. A bunch of our jobs have a pre build step that basically does a:</p>


<pre>
ps aux | grep [the-thing-you-want-to-kill] | grep -v grep | awk "{ print \$2 }"
</pre>

<p>and pass those pids to kill -9</p>

<p>That kills every "zombie/parent-less" process BEFORE the build runs.  This ensures the build will run in a relatively clean environment.<br/>
Just be careful what you kill <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>This is only a workaround, but it works for us.<br/>
Hope this helps!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dbogardus's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=247755&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-247755">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I've kept my company on Jenkins version 1.586 this whole time. It does not suffer from this bug and cleans up all surefire and phantomjs processes nicely. </p>
</pre>
</details>

</details>
<p>I've kept my company on Jenkins version 1.586 this whole time. It does not suffer from this bug and cleans up all surefire and phantomjs processes nicely. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mihelich's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=248934&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-248934">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I have a simple repro case with a freestyle project and "Execute shell" build steps. I'm on Jenkins 1.625.3, Linux master, Linux slave.</p>

<p>I hit this because we have freestyle projects where several build processes execute in parallel, but use flock to gate access to a shared resource. Sometimes, when a job is aborted, one of more of these child processes persist and prevent future jobs on the slave from acquiring the shared lock. The flock stuff below isn't necessary to repro - just sleep gives similar behavior - but it matches my use case and makes it easy to identify affected processes with fuser.</p>

<p>Create a freestyle project with two build steps:</p>
<ol>
	<li>Execute shell
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>#!/bin/bash -ex
nohup flock /var/lock/mylockfile sleep 1h &amp;
</pre>
</div></div></li>
	<li>Execute shell
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>#!/bin/bash -ex
sleep 1h
</pre>
</div></div></li>
</ol>


<p>Then abort the job (manually or by timeout). flock and its child sleep process persist, and continue to hold the lock.</p>

<p>This is the simplest project configuration I could construct. In all of these cases, the child processes are killed as expected:</p>
<ul>
	<li>Omitting the second "Execute shell."</li>
	<li>Combining them into a single "Execute shell."</li>
	<li>Failing by means other than abort, e.g. <tt>/bin/false</tt> in the second "Execute shell."</li>
</ul>


<p>Sample results below. While the job is running, the lock is in use as expected:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>$ fuser /var/lock/mylockfile
22733 22734

$ ps -p 22733,22734 -o pid,ppid,stat,lstart,args
  PID  PPID STAT                  STARTED COMMAND
22733     1 S    Wed Feb 24 00:57:51 2016 flock /var/lock/mylockfile sleep 1h
22734 22733 S    Wed Feb 24 00:57:51 2016 sleep 1h
</pre>
</div></div>

<p>Then abort the job:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>[experimental_jenkins_26048] $ /bin/bash -ex /tmp/hudson8042917752397215577.sh
+ nohup flock /var/lock/mylockfile sleep 1h
[experimental_jenkins_26048] $ /bin/bash -ex /tmp/hudson4924658810125221857.sh
+ sleep 1h
Build timed out (after 3 minutes). Marking the build as aborted.
Build was aborted
Finished: ABORTED
</pre>
</div></div>

<p>Afterwards, the processes are still alive:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>$ ps -p 22733,22734 -o pid,ppid,stat,lstart,args
  PID  PPID STAT                  STARTED COMMAND
22733     1 S    Wed Feb 24 00:57:51 2016 flock /var/lock/mylockfile sleep 1h
22734 22733 S    Wed Feb 24 00:57:51 2016 sleep 1h
</pre>
</div></div>

<p>BUILD_ID is unchanged, so ProcessTreeKiller should find them:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>$ strings /proc/22733/environ | grep BUILD_ID
BUILD_ID=17
$ strings /proc/22734/environ | grep BUILD_ID
BUILD_ID=17
</pre>
</div></div>
</pre>
</details>

</details>
<p>I have a simple repro case with a freestyle project and "Execute shell" build steps. I'm on Jenkins 1.625.3, Linux master, Linux slave.</p>

<p>I hit this because we have freestyle projects where several build processes execute in parallel, but use flock to gate access to a shared resource. Sometimes, when a job is aborted, one of more of these child processes persist and prevent future jobs on the slave from acquiring the shared lock. The flock stuff below isn't necessary to repro - just sleep gives similar behavior - but it matches my use case and makes it easy to identify affected processes with fuser.</p>

<p>Create a freestyle project with two build steps:</p>
<ol>
	<li>Execute shell


```
#!/bin/bash -ex
nohup flock /var/lock/mylockfile sleep 1h &
```</li>
	<li>Execute shell


```
#!/bin/bash -ex
sleep 1h
```</li>
</ol>


<p>Then abort the job (manually or by timeout). flock and its child sleep process persist, and continue to hold the lock.</p>

<p>This is the simplest project configuration I could construct. In all of these cases, the child processes are killed as expected:</p>
<ul>
	<li>Omitting the second "Execute shell."</li>
	<li>Combining them into a single "Execute shell."</li>
	<li>Failing by means other than abort, e.g. <tt>/bin/false</tt> in the second "Execute shell."</li>
</ul>


<p>Sample results below. While the job is running, the lock is in use as expected:</p>


```
$ fuser /var/lock/mylockfile
22733 22734

$ ps -p 22733,22734 -o pid,ppid,stat,lstart,args
  PID  PPID STAT  STARTED COMMAND
22733     1 S    Wed Feb 24 00:57:51 2016 flock /var/lock/mylockfile sleep 1h
22734 22733 S    Wed Feb 24 00:57:51 2016 sleep 1h
```

<p>Then abort the job:</p>


```
[experimental_jenkins_26048] $ /bin/bash -ex /tmp/hudson8042917752397215577.sh
+ nohup flock /var/lock/mylockfile sleep 1h
[experimental_jenkins_26048] $ /bin/bash -ex /tmp/hudson4924658810125221857.sh
+ sleep 1h
Build timed out (after 3 minutes). Marking the build as aborted.
Build was aborted
Finished: ABORTED
```

<p>Afterwards, the processes are still alive:</p>


```
$ ps -p 22733,22734 -o pid,ppid,stat,lstart,args
  PID  PPID STAT  STARTED COMMAND
22733     1 S    Wed Feb 24 00:57:51 2016 flock /var/lock/mylockfile sleep 1h
22734 22733 S    Wed Feb 24 00:57:51 2016 sleep 1h
```

<p>BUILD_ID is unchanged, so ProcessTreeKiller should find them:</p>


```
$ strings /proc/22733/environ | grep BUILD_ID
BUILD_ID=17
$ strings /proc/22734/environ | grep BUILD_ID
BUILD_ID=17
```

<!-- ### Imported Jira references for easier searching -->
<!-- [jira_issue_key=JENKINS-26048] -->
<!-- [jira_comment_id=248934] -->
<!-- [comment_author=mihelich] -->

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ssbarnea.png" title="ssbarnea's avatar" /> ssbarnea's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=262133&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-262133">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I faced the same bug with Jenkins 2.9 and some shell jobs started via pipelines. It does not always happen so I don't really know what is causing it.</p>
</pre>
</details>

</details>
<p>I faced the same bug with Jenkins 2.9 and some shell jobs started via pipelines. It does not always happen so I don't really know what is causing it.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=265395&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-265395">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Possibly related discussion: <a href="https://github.com/jenkinsci/docker/issues/54" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/docker/issues/54</a></p>
</pre>
</details>

</details>
<p>Possibly related discussion: <a href="https://github.com/jenkinsci/docker/issues/54" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/docker/issues/54</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/claudiocurzi.png" title="claudiocurzi's avatar" /> claudiocurzi's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=265733&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-265733">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Hi, I have the same problem on my windows server when build running on the slave machine.</p>

<p>Right now I have resolved by running builds criticism on the master.</p>

<p>I use Jenkins 2.11.</p>
</pre>
</details>

</details>
<p>Hi, I have the same problem on my windows server when build running on the slave machine.</p>

<p>Right now I have resolved by running builds criticism on the master.</p>

<p>I use Jenkins 2.11.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>cryssaegrim's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=266397&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-266397">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I also have seen Jenkins not killing off subprocesses when a step (regular, multiphase, and the new pipeline plugin) jobs, using Jenkins 2.15, on Windows 10 systems.  I'm happy to provide further machine specs &amp; repro.</p>

<p>My test case was running 'python -c "import time;time.sleep(100000)"'  and aborting/terminating the job (both) from Jenkins.  Using pipeline.</p>

<p>This can leave file handles open and block subsequent attempts to build in the same workspace, causing failures.</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=claudiocurzi" class="user-hover" rel="claudiocurzi">claudiocurzi</a> What is the build criticism on master you are referring to?</p>
</pre>
</details>

</details>
<p>I also have seen Jenkins not killing off subprocesses when a step (regular, multiphase, and the new pipeline plugin) jobs, using Jenkins 2.15, on Windows 10 systems.  I'm happy to provide further machine specs & repro.</p>

<p>My test case was running 'python -c "import time;time.sleep(100000)"'  and aborting/terminating the job (both) from Jenkins.  Using pipeline.</p>

<p>This can leave file handles open and block subsequent attempts to build in the same workspace, causing failures.</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=claudiocurzi" class="user-hover" rel="claudiocurzi">claudiocurzi</a> What is the build criticism on master you are referring to?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/claudiocurzi.png" title="claudiocurzi's avatar" /> claudiocurzi's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=266477&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-266477">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I have only builds using "Execute Windows Batch Command".</p>

<p>All my builds when running on the slave server and it being aborted by the timeout settings on the log, the process on the slave server continue to run.</p>
</pre>
</details>

</details>
<p>I have only builds using "Execute Windows Batch Command".</p>

<p>All my builds when running on the slave server and it being aborted by the timeout settings on the log, the process on the slave server continue to run.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ppaczyn.png" title="ppaczyn's avatar" /> ppaczyn's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=270510&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-270510">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>For us it seems to work fine when jobs are run on master, but processes are not killed when job runs on slave (Maven job type).</p>
</pre>
</details>

</details>
<p>For us it seems to work fine when jobs are run on master, but processes are not killed when job runs on slave (Maven job type).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jchatham's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=272104&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-272104">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I recently encountered this issue in the context of using the build-timeout plugin to stop a maven build; I don't have an actual fix for the bug in the maven module, but was able to put together some code to make the build timeout plugin kill off appropriate child processes.  See my comment on <a href="https://issues.jenkins.io/browse/JENKINS-28125" title="No e-mail is sent if build is aborted" class="issue-link" data-issue-key="JENKINS-28125"><del>JENKINS-28125</del></a> if you think that code might be of use to you.</p>
</pre>
</details>

</details>
<p>I recently encountered this issue in the context of using the build-timeout plugin to stop a maven build; I don't have an actual fix for the bug in the maven module, but was able to put together some code to make the build timeout plugin kill off appropriate child processes.  See my comment on <a href="https://issues.jenkins.io/browse/JENKINS-28125" title="No e-mail is sent if build is aborted" class="issue-link" data-issue-key="JENKINS-28125"><del>JENKINS-28125</del></a> if you think that code might be of use to you.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>tib's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=278092&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-278092">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>We are seeing this issue.
Linux master, Windows slaves.
No Maven. Just regular pipeline jobs, starting a background process like this:</p>

<p>bat "start MyProc.exe"</p>

<p>From my understanding the process should be killed when the job is done? But it keeps running.</p>
</pre>
</details>

</details>
<p>We are seeing this issue.<br/>
Linux master, Windows slaves.<br/>
No Maven. Just regular pipeline jobs, starting a background process like this:</p>

<p>bat "start MyProc.exe"</p>

<p>From my understanding the process should be killed when the job is done? But it keeps running.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=284750&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284750">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tib" class="user-hover" rel="tib">tib</a> currently Pipeline <tt>sh</tt>/<tt>bat</tt> steps kill all processes when the build is interrupted, but not when the main process simply exits on its own. Could be filed under <tt>durable-task-plugin</tt>. Unrelated to this issue.</p>
</pre>
</details>

</details>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tib" class="user-hover" rel="tib">tib</a> currently Pipeline <tt>sh</tt>/<tt>bat</tt> steps kill all processes when the build is interrupted, but not when the main process simply exits on its own. Could be filed under <tt>durable-task-plugin</tt>. Unrelated to this issue.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> oleg_nenashev's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=298286&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-298286">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Just a summary of more or less recent changes on this front, in Windows:</p>
<ul>
	<li>In Jenkins 2.34 (WinP 1.24) there were fixes of the process management logic in Windows: <a href="https://issues.jenkins.io/browse/JENKINS-20913" title="[jar caching] - winp.x64.dll access conflicts" class="issue-link" data-issue-key="JENKINS-20913"><del>JENKINS-20913</del></a> and <a href="https://issues.jenkins.io/browse/JENKINS-24453" title="Jenkins server got blue screen to death" class="issue-link" data-issue-key="JENKINS-24453"><del>JENKINS-24453</del></a>. Before this version process termination is not reliable</li>
	<li>In Jenkins 2.50+ (WinSW 2.0) there were many changes improving the process termination logic if the service gets terminated/restarted. Changelog: <a href="https://github.com/kohsuke/winsw/blob/master/CHANGELOG.md" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/kohsuke/winsw/blob/master/CHANGELOG.md</a> </li>
	<li>In Jenkins 2.50+ (WinSW 2.0) I have added the Runaway Process Killer extension, which kills spawned processes if Windows Service Wrapper executable gets aborted. In some cases it may also cause build processes to runaway</li>
</ul>


<p>What is known to NOT work:</p>
<ul>
	<li>Termination of 64bit processes in Windows if your Jenkins master/agent runs on 32bit JRE/JDK. Won't fix though it needs to be documented somewhere</li>
</ul>


<p>I feel for Windows the issue is more or less resolved. Or it needs retesting at least. </p>

<p>Regarding Unix OS, we need to understand if it still happens in modern Jenkins versions.</p>
</pre>
</details>

</details>
<p>Just a summary of more or less recent changes on this front, in Windows:</p>
<ul>
	<li>In Jenkins 2.34 (WinP 1.24) there were fixes of the process management logic in Windows: <a href="https://issues.jenkins.io/browse/JENKINS-20913" title="[jar caching] - winp.x64.dll access conflicts" class="issue-link" data-issue-key="JENKINS-20913"><del>JENKINS-20913</del></a> and <a href="https://issues.jenkins.io/browse/JENKINS-24453" title="Jenkins server got blue screen to death" class="issue-link" data-issue-key="JENKINS-24453"><del>JENKINS-24453</del></a>. Before this version process termination is not reliable</li>
	<li>In Jenkins 2.50+ (WinSW 2.0) there were many changes improving the process termination logic if the service gets terminated/restarted. Changelog: <a href="https://github.com/kohsuke/winsw/blob/master/CHANGELOG.md" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/kohsuke/winsw/blob/master/CHANGELOG.md</a> </li>
	<li>In Jenkins 2.50+ (WinSW 2.0) I have added the Runaway Process Killer extension, which kills spawned processes if Windows Service Wrapper executable gets aborted. In some cases it may also cause build processes to runaway</li>
</ul>


<p>What is known to NOT work:</p>
<ul>
	<li>Termination of 64bit processes in Windows if your Jenkins master/agent runs on 32bit JRE/JDK. Won't fix though it needs to be documented somewhere</li>
</ul>


<p>I feel for Windows the issue is more or less resolved. Or it needs retesting at least. </p>

<p>Regarding Unix OS, we need to understand if it still happens in modern Jenkins versions.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> oleg_nenashev's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=303539&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-303539">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I am closing the issue according to the comment above. If you still experience the issue, please feel free to create follow-up issues for your cases.</p>
</pre>
</details>

</details>
<p>I am closing the issue according to the comment above. If you still experience the issue, please feel free to create follow-up issues for your cases.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dbogardus's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=304164&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304164">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Verified this is still happening in latest jenkins version on CentOS 6. </p>

<p>For the test I'm running java tests with the failsafe plugin, which launches multiple jvms for concurrent execution and many phantomjs processes. </p>

<p>1.586 - When build is stopped, all child processes are stopped</p>

<p>1.587 - When build is stopped, all child processes keep running</p>

<p>2.66.1 - When build is stopped, all child processes keep running</p>

<p> </p>
</pre>
</details>

</details>
<p>Verified this is still happening in latest jenkins version on CentOS 6. </p>

<p>For the test I'm running java tests with the failsafe plugin, which launches multiple jvms for concurrent execution and many phantomjs processes. </p>

<p>1.586 - When build is stopped, all child processes are stopped</p>

<p>1.587 - When build is stopped, all child processes keep running</p>

<p>2.66.1 - When build is stopped, all child processes keep running</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dbogardus's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=304170&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304170">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck" class="user-hover" rel="danielbeck">danielbeck</a> , I used the latest weekly version - 2.66.1 . (It's just what came down from yum)</p>

<p>I am able to reproduce it with the simple java/mvn/surefire scenario created by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rddesmond" class="user-hover" rel="rddesmond">rddesmond</a> :</p>

<p><a href="https://issues.jenkins-ci.org/browse/JENKINS-28968?focusedCommentId=230600&amp;page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-230600" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-28968?focusedCommentId=230600&amp;page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-230600</a></p>

<p>With this small example, 1.586 cleans up the surefire process, and 2.66.1 does not. </p>

<p> </p>

<p> </p>
</pre>
</details>

</details>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck" class="user-hover" rel="danielbeck">danielbeck</a> , I used the latest weekly version - 2.66.1 . (It's just what came down from yum)</p>

<p>I am able to reproduce it with the simple java/mvn/surefire scenario created by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rddesmond" class="user-hover" rel="rddesmond">rddesmond</a> :</p>

<p><a href="https://issues.jenkins-ci.org/browse/JENKINS-28968?focusedCommentId=230600&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-230600" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-28968?focusedCommentId=230600&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-230600</a></p>

<p>With this small example, 1.586 cleans up the surefire process, and 2.66.1 does not. </p>

<p> </p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> oleg_nenashev's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=304233&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304233">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dbogardus" class="user-hover" rel="dbogardus">dbogardus</a> I propose to move the discussion back to the <a href="https://issues.jenkins.io/browse/JENKINS-28968" title="Aborting builds does not kill surefire sub-process" class="issue-link" data-issue-key="JENKINS-28968">JENKINS-28968</a> thread. Maven Project plugin is a very specific case in terms of the architecture, so I would rather vote for handling it separately until we confirm there is a generic issue in the core. So far I am not conviced</p>
</pre>
</details>

</details>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dbogardus" class="user-hover" rel="dbogardus">dbogardus</a> I propose to move the discussion back to the <a href="https://issues.jenkins.io/browse/JENKINS-28968" title="Aborting builds does not kill surefire sub-process" class="issue-link" data-issue-key="JENKINS-28968">JENKINS-28968</a> thread. Maven Project plugin is a very specific case in terms of the architecture, so I would rather vote for handling it separately until we confirm there is a generic issue in the core. So far I am not conviced</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dbogardus's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=304280&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304280">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev" class="user-hover" rel="oleg_nenashev">oleg_nenashev</a> Acknowledged. I did try a few things to reproduce the bug outside of the maven/surefire scenario but was unable to. </p>
</pre>
</details>

</details>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev" class="user-hover" rel="oleg_nenashev">oleg_nenashev</a> Acknowledged. I did try a few things to reproduce the bug outside of the maven/surefire scenario but was unable to. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>pyrocks's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=324599&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-324599">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Hi,</p>

<p>We're facing a similar issue - but we're using gradle and not using maven or surefire.</p>

<p>We use pipeline and call gradle using the bat step (windows 64 bit slaves, master is on windows too).</p>

<p>This happens on 2 separate projects in which gradle spawns child processes for testing using various other tools. Sometimes 'gradle test' hangs due to bugs in the test code/scenarios and when we stop the build - there are always java processes left up which we have to kill manually.</p>

<p>Verified that ProcessTreeKiller is not disabled.</p>

<p>Using Jenkins 2.73.3, pipeline suite 2.5.</p>
</pre>
</details>

</details>
<p>Hi,</p>

<p>We're facing a similar issue - but we're using gradle and not using maven or surefire.</p>

<p>We use pipeline and call gradle using the bat step (windows 64 bit slaves, master is on windows too).</p>

<p>This happens on 2 separate projects in which gradle spawns child processes for testing using various other tools. Sometimes 'gradle test' hangs due to bugs in the test code/scenarios and when we stop the build - there are always java processes left up which we have to kill manually.</p>

<p>Verified that ProcessTreeKiller is not disabled.</p>

<p>Using Jenkins 2.73.3, pipeline suite 2.5.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> oleg_nenashev's <a href="https://issues.jenkins.io/browse/JENKINS-26048?focusedId=356874&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-356874">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I will have no time to work on it anytime soon, please see <a href="https://groups.google.com/d/msg/jenkinsci-dev/uc6NsMoCFQI/AIO4WG1UCwAJ" class="external-link" target="_blank" rel="nofollow noopener">https://groups.google.com/d/msg/jenkinsci-dev/uc6NsMoCFQI/AIO4WG1UCwAJ</a> for the context. I will unassign it so that somebody else can work on it</p>
</pre>
</details>

</details>
<p>I will have no time to work on it anytime soon, please see <a href="https://groups.google.com/d/msg/jenkinsci-dev/uc6NsMoCFQI/AIO4WG1UCwAJ" class="external-link" target="_blank" rel="nofollow noopener">https://groups.google.com/d/msg/jenkinsci-dev/uc6NsMoCFQI/AIO4WG1UCwAJ</a> for the context. I will unassign it so that somebody else can work on it</p>
